### PR TITLE
fix: reset loss logging after checkpoint resume

### LINKED ISSUE #6237: [BUG] Loss logging double-counts the first iteration after checkpoint resume
## Description

When training resumes from a checkpoint, the first iteration of the resumed process is treated as a fresh-run first iteration by `training_log`. This causes the loss accumulator to retain the first resumed loss and add it to the next iteration.

The issue is visible with `--log-interval 1`. It affects reported loss/timing statistics, but not the forward computation, gradients, optimizer update, or checkpoint contents.

## Environment

- Repository: `NVIDIA/Megatron-LM`
- Branch: `core_r0.17.0`
- Commit: `963bf39218e8bb83a1203b40293358498322be50`
- BF16 distributed training
- Observed on an Ascend/NPU run; the affected code path appears device-independent
- `--log-interval 1`
- Checkpoint saved at iteration 50, then resumed at iteration 51

## Reproduction

1. Run a deterministic BF16 training job through iteration 53 with `--save-interval 50` and `--log-interval 1`.
2. Run the same job from the beginning and separately resume from the iteration-50 checkpoint.
3. Compare the loss logs for iterations 51--53.

The relevant control flow is:

```python
# train()
is_first_iteration = True

# training_log()
should_reset = not is_first_iteration

if iteration % args.log_interval == 0 or is_first_iteration:
    ...
    if should_reset:
        total_loss_dict[key] = torch.tensor([0.0], ...)
```

On resume, `is_first_iteration` is reset to `True`, while `args.iteration > 0` and `total_loss_dict` starts empty.

## Observed behavior

At iteration 52:

- uninterrupted run: `lm loss = 11.156137466430664`, `advanced iterations = 1`
- resumed run: accumulated loss `= 22.273590087890625`, `advanced iterations = 2`, average `= 11.136795043945312`

Iteration 53 becomes normal again because iteration 52 resets the accumulator. Additional boundary diagnostics show that the batch, local/global loss, and pre-optimizer gradient hashes match between uninterrupted and resumed runs; only the `loss_log` accumulator differs.

## Expected behavior

After resuming, `--log-interval 1` should report the loss for the current iteration only, matching an uninterrupted run. The first resumed iteration should not be retained and double-counted in the next log interval.

## Possible fix

A minimal fix is to preserve the current first-iteration accumulation behavior only for a fresh run:

```python
should_reset = not is_first_iteration or args.iteration > 0
```

Alternatively, the first-iteration logging state could be initialized from the restored global iteration rather than process lifetime.

### PR DESCRIPTION
**Pull request by Burceoneo12345**
## What does this PR do?

Fixes loss logging after checkpoint resume when `--log-interval 1` is used.

The first iteration of a resumed process is currently treated as the first iteration of a fresh training run. This keeps its loss in `total_loss_dict`, so the next iteration logs the sum of two losses while dividing by two.

## Why is this needed?

For a fresh run, retaining the first logged loss is intentional because the first iteration is logged before the normal logging interval.

For a resumed run, `total_loss_dict` starts empty while `args.iteration > 0`. Retaining the first resumed loss causes it to be double-counted in the next log. In the observed case:

- uninterrupted iteration 52: `lm loss = 11.156137466430664`
- resumed iteration 52: accumulated loss `= 22.273590087890625`, `advanced iterations = 2`, average `= 11.136795043945312`

The forward loss and pre-optimizer gradient diagnostics match; this change only corrects the logging accumulator boundary.

## What changes?

Use independent accumulation boundaries for loss statistics and timers:

- Reset loss statistics after the first resumed iteration.
- Preserve timer accumulation through that iteration.
- Track timer iterations separately so elapsed time and operation timers are normalized by the exact number of accumulated timer samples.

## Validation

- Added unit coverage for a fresh first iteration and resumed first iterations with `log_interval` values 1 and 10.
- Added sequential resume coverage for iterations 51 through 60, verifying that timer values use 10 samples while loss statistics use the 9 iterations after the forced first log.
- `git diff --check`
- `PYTHONPYCACHEPREFIX=/tmp/megatron-pr-resume-loss-pycache python -m py_compile megatron/training/training.py tests/unit_tests/test_training.py`
- `uvx ruff check megatron/training/training.py tests/unit_tests/test_training.py`
- `uvx --from isort isort --check-only tests/unit_tests/test_training.py`
- `uvx --from black==24.10.0 black --check tests/unit_tests/test_training.py`
- The distributed GPU unit test was not run on this local macOS host.
- NPU runtime validation was not rerun locally.

Fixes #6237

### CODE REVIEW

File: megatron/training/training.py [hunk 1]
```diff
@@ -1967,8 +1967,10 @@ def training_log(
     one_logger = get_one_logger()
     energy_monitor = get_energy_monitor()
 
-    # On first iteration, log stats but don't reset accumulators so normal interval stats remain accurate.
-    should_reset = not is_first_iteration
+    # Preserve the first-iteration accumulation only for a fresh training run. A resumed
+    # process starts with an empty total_loss_dict, so keeping the first resumed loss would
+    # average it with the next iteration.
+    should_reset = not is_first_iteration or args.iteration > 0
```
[the remark this answers is not in the document]
Comment (Burceoneo12345) [reply megatron/training/training.py]: Addressed in b87ebb2f3. The loss-stat reset is now independent from the timer reset: resumed first iterations clear loss/counter accumulators, while interval and operation timers continue accumulating until the normal logging boundary. Added unit coverage for fresh start, resume with log_interval=1, and resume with log_interval=10.

**Comment by Burceoneo12345** [comment]
@NVIDIA/mcore-oncall This PR has been updated to address the timer-accumulation reviews, includes regression coverage for fresh and resumed first iterations, is rebased onto the latest core_r0.17.0, and passes DCO.

Could a maintainer please vet the external-contributor workflows and add the Run functional tests label? GitHub does not allow the PR author to apply that label, and the PR now adds a new regression test case.

Please also confirm whether this fix should target core_r0.17.0 directly or land on main first and be backported; the same first-iteration reset logic is still present on main.

File: megatron/training/training.py [hunk 2]
```diff
@@ -1967,8 +1967,11 @@ def training_log(
     one_logger = get_one_logger()
     energy_monitor = get_energy_monitor()
 
-    # On first iteration, log stats but don't reset accumulators so normal interval stats remain accurate.
-    should_reset = not is_first_iteration
+    # Preserve first-iteration timer accumulation for the next normal logging interval.
+    # Loss stats are preserved only for a fresh run; resumed runs start with an empty
+    # total_loss_dict, so keeping the first resumed loss would count it twice.
+    should_reset_timers = not is_first_iteration
+    should_reset_loss_stats = should_reset_timers or args.iteration > 0
```
[the remark this answers is not in the document]
Comment (Burceoneo12345) [reply megatron/training/training.py:1974]: Addressed in 25e4d007e. Loss and timer accumulation now have independent iteration counters. For a resume at 50 with log_interval=10, the forced iteration-51 log resets only the loss counters; at iteration 60, elapsed time and operation timers are normalized by 10 timer iterations, while the loss average remains normalized by the 9 post-reset loss iterations. The regression test now executes iterations 51 through 60 and checks both averages.

**Comment by Burceoneo12345** [comment]
Hi @hexinw-nvidia, @NVIDIA/training-adlr, and @NVIDIA/training-nemo — gentle ping on this PR, which has been waiting for maintainer review.

It fixes a checkpoint-resume edge case in the first-iteration loss/timer accumulator logic introduced in #2954 and adds focused regression coverage. Could one of the relevant owners please review it, vet the external-contributor workflows/add the appropriate test label, and advise whether it should target `core_r0.17.0` directly or land on `main` first?

Thank you!

**Comment by Burceoneo12345** [comment]
Hi @hexinw-nvidia and @NVIDIA/ci, following up on this PR: the [Review Trigger run](https://github.com/NVIDIA/Megatron-LM/actions/runs/32226053574) has now expired with the annotation: "This workflow run required approval but was not approved before it expired." No test jobs started.

Could you please help vet the external-contributor workflows and re-trigger/approve the run, or direct this to the maintainer who can? @hexinw-nvidia, could you also take a look at the checkpoint-resume logging fix related to #2954 when you have a chance?

Thank you!
